app/controllers/authenticate_controller.rb

```diff
@@ -77,7 +77,16 @@ def handle_authentication_error(err)
     err.backtrace.each do |line|
       logger.debug(line)
     end
-    raise Unauthorized
+
+    case err
+    when Conjur::RequiredResourceMissing
+    when Conjur::RequiredSecretMissing
+    when Authentication::Security::ServiceNotDefined
+    when Authentication::Security::NotWhitelisted
+      raise NotImplemented, err.message
+    else
+      raise Unauthorized
+    end
```

Comment:
This belongs in the relevant controllers.  Didn't look into the first two, but `Authentication::Security::ServiceNotDefined` and `Authentication::Security::NotWhitelisted` belong in the `AuthenticateController` -- they're not application wide concerns.

Reply:
I don't understand. This code is in `authenticate_controller` as you say

Reply:
what I'm trying to accomplish here is that in cases where we have an error in the authn flow, and we need further user configuration then we'll raise a NotImplemented error which will cause a 501 response code